### ISSUE TITLE
SQLite support

### ISSUE THREAD

**Issue by stephenh** [stephenh/joist-ts#467]
Just tracking what would be necessary, after a very small spike:

- Big: No [pg-structure](https://www.pg-structure.com/) for SQLite; sqlite doesn't even support `information_schema`, it has a `sqlite_master` that just as the _raw DDL_ for each object, which would then need to be parsed :-|
  - Instead of trying to reflect out the schema, try the Prisma approach (and even it's literal file format) of declaring the model outside the db
- Medium: No `SEQUENCE`s, so would need to do INSERT+UPDATEs
- Good: Does support deferred FK constraints

**Comment by stephenh** [comment]
Ah nice, hey @AustinGil !

Kinda surprised, I did some googling after seeing your comment--one of the blockers here was lack of sequences in sqlite, but turns out sqlite's `autoincrement` primary keys have an internal [sqlite_sequence](https://www.sqlite.org/fileformat2.html#seqtab) table that can very likely be used (carefully, pending more research) to mimic Joist's up-front/pre-INSERT assignment of entity ids, which was one of the pg-isms that would be ~not-hard-but-annoying to move away from:

```sql
sqlite> create table bar (id integer primary key autoincrement);
sqlite> insert into bar (id) values (null);
sqlite> select * from bar;
1
sqlite> select * from sqlite_sequence;
bar|1
-- pretend joist bulk pre-assigns 10 new ids for bar entities 
sqlite> update sqlite_sequence set seq = 10;
sqlite> insert into bar (id) values (null);
sqlite> select * from bar;
1
11
```

So pulling the metadata from the SQLite schema is probably the biggest blocker at this point.

**Comment by AustinGil** [comment]
Yeah, I wish I knew more about db adapters so I could be of service, but I'm more of a frontend guy. Trying to find an ORM I really like, and hopefully one that works with SQLite. I listened to the JS Party on this one and it sounds great.

**Comment by stephenh** [comment]
Np! Especially for the first one or two "not postgres" adapters, it will likely be something myself/one of our core maintainers will need to do, just to even get Joist's adapter layer setup well to really do "more than just postgres".

I did some more googling and really seems like the blocker is finding a "pg-structure but for sqlite"--like this `sqlite_master` table will tell you all of the tables in the database, but to get the columns / foreign keys / etc., you have to parse the `CREATE TABLE` DDL. Which just doesn't sound like fun... And I haven't found anything on npm that will do it for us.
